### PR TITLE
API Only expose public extension methods

### LINKED ISSUE #6210: RFC Extensions must whitelist provided methods
Extensions currently allow method extensions in the following ways: Either by methods that can be invoked via `$obj->extend('method')`, or by providing a default method implementation that can be invoked if the parent object does not have a definition for `$obj->method()`.

However there is no distinction between the two forms of methods, meaning that it's impossible to provide an extension point without also providing a default method implementation. This can be hazardous for extensions which are ONLY intended to work with `->extend` and should not increase the parent object's surface area.

I propose that we add `provide_methods` config to extensions. By default methods on extensions can only be invoked via `->extend`. However, a default method can also be provided, given that it's registered.

Additionally, the `provides_methods` can support aliases, so you can have two different method implementations.

E.g. Versioned could look like the below:

``` php


class Versioned extends Extension {
    private static $provides_methods = [
        'canArchive' => 'defaultCanArchive'
    ];


    /**
     * Can be invoked on parent object via $obj->canArchive()
     */
    public function defaultCanArchive($member = null) {
        // Default method implementation
        return $this->extendedCan('canArchive');
    }

    /**
     * can only be invoked via $obj->extend('canArchive')
     */
    public function canArchive($method = null) {
        return null;
    }
}

```

provides_methods could also support non-associative values, which are a 1-to-1 mapping.

A downside to this I can see is that it breaks php type and method hinting, which previously was supported via `@mixin Versioned`, so some more thought is necessary here. :)

# PRs created:

 - [ ] https://github.com/silverstripe/silverstripe-framework/pull/7752
 - [ ] https://github.com/silverstripe/silverstripe-versioned/pull/84
 - [ ] https://github.com/silverstripe/silverstripe-asset-admin/pull/723
 - [x] https://github.com/silverstripe/silverstripe-testsession/pull/55

### PR DESCRIPTION
**Pull request by tractorcow**
Built on top of https://github.com/silverstripe/silverstripe-framework/pull/6847, replaces it.

Fixes #6210 using `extend*` solution for extension methods.

I'll write some more docs and tests, but creating PR to collect feedback.

extended methods can be protected, but now all "default" methods added by extensions must be public. See `CustomMethods::findBuiltInMethods` which filters only public methods for exposure.

I am wondering if we require `->extend` methods to be protected? That way all extension methods are either added to the parent object as an extra method, or can be used with `->extend`.

Also see  https://github.com/silverstripe/silverstripe-versioned/pull/84

### CODE REVIEW

File: src/Core/CustomMethods.php [hunk 1]
```diff
@@ -157,66 +159,87 @@ public function hasMethod($method)
...
+        $lowerClass = strtolower(static::class);
+        if (!isset(self::$extra_methods[$lowerClass])) {
             $this->defineMethods();
         }
 
-        if (isset(self::$extra_methods[static::class][strtolower($method)])) {
-            return self::$extra_methods[static::class][strtolower($method)];
-        }
-        return null;
+        return self::$extra_methods[$lowerClass][strtolower($method)] ?? null;
```
[the remark this answers is not in the document]
Comment (robbieaverill) [reply src/Core/CustomMethods.php]: It’s a beautiful new PHP7 addition :-D

[the remark this answers is not in the document]
Comment (dhensby) [reply src/Core/CustomMethods.php]: Shocker! Docs?

[the remark this answers is not in the document]
Comment (tractorcow) [reply src/Core/CustomMethods.php]: It's master so it can be 7.1 only API.

[the remark this answers is not in the document]
Comment (tractorcow) [reply src/Core/CustomMethods.php]: Docs https://secure.php.net/manual/en/migration70.new-features.php#migration70.new-features.null-coalesce-op

[the remark this answers is not in the document]
Comment (dhensby) [reply src/Core/CustomMethods.php]: nice, so it's a bit like doing `||` in JS

[the remark this answers is not in the document]
Comment (tractorcow) [reply src/Core/CustomMethods.php]: Yes except it won't error if those array keys aren't set; Not having to duplicate the whole expression saves a lot of pain.

**Comment by tractorcow** [comment]
@dhensby why didn't you merge this? The tests look fine to me.


image

**Comment by tractorcow** [comment]
Tests fixed with https://github.com/silverstripe/silverstripe-testsession/pull/55; Needs a tag to get it running with tests though.

**Comment by dhensby** [comment]
Shall we remove prefer stable for 5.x branch?

**Comment by dhensby** [comment]
@tractorcow looks like a tag wasn't needed, it's pulling out `dev-master` for testsession.

**Comment by dhensby** [comment]
some 4.x changes also needed merging into master (public webroot work) so I've done that too

**Comment by tractorcow** [comment]
Given this a rebase. ;)

**Comment by tractorcow** [comment]
@dhensby most unit tests are green, but we need to get some of the other modules migrated to get everything green. What do you think about giving this another look?

**Comment by dhensby** [comment]
@tractorcow I'm happy with this - seems the test failures are really around behat and the unit tests seem fine(??)

**Comment by tractorcow** [comment]
Yeah it'll probably need the other modules merged before behat works.

**Comment by tractorcow** [comment]
Notably, canPublish() won't work until https://github.com/silverstripe/silverstripe-versioned/pull/84 is merged (since the func_num_arg() hack no longer works).

**Comment by dhensby** [comment]
@tractorcow so there's a circular dependency on these two PRs? will builds be green if I merge both?

**Comment by tractorcow** [comment]
Yep

**Comment by dhensby** [comment]
@tractorcow looks like they aren't green :(

**Comment by tractorcow** [comment]
Fixes:
  - https://github.com/silverstripe/silverstripe-cms/pull/2103
  - https://github.com/silverstripe/silverstripe-framework/pull/7879

**Pull request merged**